lib/ecto/changeset.ex

```diff
@@ -225,33 +225,86 @@ defmodule Ecto.Changeset do
                optional: optional}
   end
 
-  defp process_param(key, kind, params, types, model, {changes, errors}) do
+  defp process_param({key, fun}, kind, params, types, model, acc) do
     {key, param_key} = cast_key(key)
+    type = type!(types, key, fun)
     current = Map.get(model, key)
-    type = type!(types, key)
 
+    do_process_param(key, param_key, kind, params, type, current, acc)
+  end
+
+  defp process_param(key, kind, params, types, model, acc) do
+    {key, param_key} = cast_key(key)
+    type = type!(types, key, nil)
+    current = Map.get(model, key)
+
+    do_process_param(key, param_key, kind, params, type, current, acc)
+  end
+
+  defp do_process_param(key, param_key, kind, params, {:embed, embed}, current, {changes, errors}) do
+
+    {key,
+     case cast_embed(param_key, embed, params, current) do
+       :missing ->
+         {changes, error_on_nil(kind, key, current, errors)}
+       {embed_changes, embed_errors} ->
+         {merge_embed(changes, key, embed_changes, &Map.put/3),
```

Comment:
I think this is too complicated?

    {Map.update(changes, key, embed_changes, &Map.merge(&1, embed_changes)),
     Keyword.update(errors, key, embed_errors, & &1 ++ embed_errors)}

Or something of sorts. You could also just override whatever is in key, I don't think we would validate the same field twice:

    {Map.put(changes, key, embed_changes),
     [{key, embed_errors}|errors]}